output/badger/output.go

```diff
@@ -0,0 +1,202 @@
+package badger
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/cube2222/octosql"
+	"github.com/cube2222/octosql/execution"
+	"github.com/cube2222/octosql/streaming/storage"
+)
+
+type Output struct {
+	EventTimeField octosql.VariableName
+}
+
+var recordsPrefix = []byte("$records$")
+var watermarkPrefix = []byte("$watermark$")
+var errorPrefix = []byte("$error$")
+var endOfStreamPrefix = []byte("$end_of_stream$")
+
+func (o *Output) ReadyForMore(ctx context.Context, tx storage.StateTransaction) error {
+	return nil
+}
+
+func (o *Output) AddRecord(ctx context.Context, tx storage.StateTransaction, inputIndex int, record *execution.Record) error {
+	if inputIndex != 0 {
+		return errors.Errorf("input with index other than 0 not allowed in output, got %d", inputIndex)
+	}
+	records := storage.NewMap(tx.WithPrefix(recordsPrefix))
+
+	recordKV := map[string]octosql.Value{}
+	for _, field := range record.Fields() {
+		recordKV[field.Name.String()] = record.Value(field.Name)
+	}
+
+	key := octosql.MakeObject(recordKV)
+
+	var octoIDs octosql.Value
+	err := records.Get(&key, &octoIDs)
+	if err == storage.ErrNotFound {
+		octoIDs = octosql.MakeTuple(nil)
+	} else if err != nil {
+		return errors.Wrap(err, "couldn't get current IDs for record value")
+	}
+	IDs := octoIDs.AsSlice()
+
+	if !record.IsUndo() {
+		newIDs := octosql.MakeTuple(append(IDs, octosql.MakeString(record.ID().Show())))
+		if err := records.Set(&key, &newIDs); err != nil {
+			return errors.Wrap(err, "couldn't add record ID to output records")
+		}
+	} else {
+		IDs = IDs[:len(IDs)-1]
```

Comment:
I don't think I understand, how do You know that ID of current record (with undo) is the last one in IDs slice?

Reply:
undo's have different ID's than the original records. So we just delete any one of them.